apps/emqx/src/emqx_channel.erl

```diff
@@ -483,60 +483,39 @@ handle_in(
-                true ->
-                    handle_out(disconnect, ?RC_NOT_AUTHORIZED, Channel);
-                false ->
-                    TopicFilters2 = [
-                        TopicFilter
-                     || {TopicFilter, ?RC_SUCCESS} <- TupleTopicFilters0
-                    ],
-                    TopicFilters3 = run_hooks(
-                        'client.subscribe',
-                        [ClientInfo, Properties],
-                        TopicFilters2
-                    ),
-                    {TupleTopicFilters1, NChannel} = process_subscribe(
-                        TopicFilters3,
-                        Properties,
-                        Channel
-                    ),
-                    TupleTopicFilters2 =
-                        lists:foldl(
-                            fun
-                                ({{Topic, Opts = #{deny_subscription := true}}, _QoS}, Acc) ->
-                                    Key = {Topic, maps:without([deny_subscription], Opts)},
-                                    lists:keyreplace(Key, 1, Acc, {Key, ?RC_UNSPECIFIED_ERROR});
-                                (Tuple = {Key, _Value}, Acc) ->
-                                    lists:keyreplace(Key, 1, Acc, Tuple)
-                            end,
-                            TupleTopicFilters0,
-                            TupleTopicFilters1
-                        ),
-                    ReasonCodes2 = [
-                        ReasonCode
-                     || {_TopicFilter, ReasonCode} <- TupleTopicFilters2
-                    ],
-                    handle_out(suback, {PacketId, ReasonCodes2}, NChannel)
-            end;
-        {error, ReasonCode} ->
-            handle_out(disconnect, ReasonCode, Channel)
+handle_in(SubPkt = ?SUBSCRIBE_PACKET(PacketId, _Properties, _TopicFilters0), Channel0) ->
+    Pipe = pipeline(
+        [
+            fun check_subscribe/2,
+            fun enrich_subscribe/2,
+            %% should never check mountpoint
+            fun check_sub_authzs/2,
+            fun check_sub_caps/2
+            %% fun run_sub_hooks/2
+        ],
+        SubPkt,
+        Channel0
+    ),
+    case Pipe of
+        {ok, NPkt = ?SUBSCRIBE_PACKET(_PacketId, CheckedTopicFilters), Channel} ->
+            %% Q: PacketId should not be changed by anywhere
+            %% even hook funcs, e.g. exhook
+
+            %% TODO:
+            %% move run_sub_hooks/2 back to pipeline after it fixed to
+            %% allow `client.subscribe` hook to reject subscriptions by reason code.
+            %% And then we could do sub for each topic filter in process_subscribe/2
+            %% to make sure each reason code is correct.
+            NTopicFilters = run_sub_hooks(NPkt, Channel),
+            {TopicFiltersWithRC, NChannel} = process_subscribe(NTopicFilters, Channel),
+            ReasonCodes2 = gen_reason_codes(CheckedTopicFilters, TopicFiltersWithRC),
+            %% END TODO
```

Comment:
Should allow `client.subscribe` hook to modify the `ReasonCode` to allow a subscription denied by authz?
or customizing `ReasonCode` by theirself (e.g. ExHook)